runtime/src/message_processor.rs

```diff
@@ -301,24 +323,167 @@ mod tests {
 
     #[test]
     fn test_verify_instruction_change_data() {
-        fn change_data(program_id: &Pubkey) -> Result<(), InstructionError> {
+        fn change_data(program_id: &Pubkey, is_credit_debit: bool) -> Result<(), InstructionError> {
             let alice_program_id = Pubkey::new_rand();
             let account = Account::new(0, 0, &alice_program_id);
-            verify_instruction(&program_id, &alice_program_id, 0, &[42], &account)
+            verify_instruction(
+                is_credit_debit,
+                &program_id,
+                &alice_program_id,
+                0,
+                &[42],
+                &account,
+            )
         }
 
         let system_program_id = system_program::id();
         let mallory_program_id = Pubkey::new_rand();
 
         assert_eq!(
-            change_data(&system_program_id),
+            change_data(&system_program_id, true),
             Ok(()),
             "system program should be able to change the data"
         );
         assert_eq!(
-            change_data(&mallory_program_id),
+            change_data(&mallory_program_id, true),
             Err(InstructionError::ExternalAccountDataModified),
             "malicious Mallory should not be able to change the account data"
         );
+
+        assert_eq!(
+            change_data(&system_program_id, false),
+            Err(InstructionError::CreditOnlyDataModified),
+            "system program should not be able to change the data if credit-only"
+        );
+    }
+
+    #[test]
+    fn test_verify_instruction_credit_only() {
+        let alice_program_id = Pubkey::new_rand();
+        let account = Account::new(0, 0, &alice_program_id);
+        assert_eq!(
+            verify_instruction(
+                false,
+                &system_program::id(),
+                &alice_program_id,
+                42,
+                &[],
+                &account
+            ),
+            Err(InstructionError::ExternalAccountLamportSpend),
+            "debit should fail, even if system program"
+        );
+    }
+
+    #[test]
+    fn test_process_message_credit_only_handling() {
+        use solana_sdk::instruction::{AccountMeta, Instruction, InstructionError};
```

Comment:
We don't use this pattern anywhere else, do we? Can you move these imports up to the top of `mod tests`?